### ISSUE TITLE
Consider hillshade-opacity property

### ISSUE THREAD

**Issue by stevage** [mapbox/mapbox-gl-js#5935]
It's a bit surprising that there is no `hillshade-opacity` property. I think it's the only layer type that doesn't have one. Currently the workaround is to use `hsla` or `rgba` colors with the alpha channel.

OTOH, maybe that becomes confusing, because I guess properties like `hillshade-shadow-color` are implicitly mixing the given color in varying degrees of alpha across the layer.

**Comment by mollymerp** [comment]
yes, it is a bit confusing, but opacity is determined by the `hillshade-*-color`  as well as the lighting angle/intensity and the slope (which vary pixel to pixel in any given layer), so if we did allow a `hillshade-opacity` property, it would behave differently than the `-opacity` properties on other layers. 

@nickidlugash @ajashton @kkaefer do you have any thoughts on this?

**Comment by kkaefer** [comment]
Hm, we could add this property for consistency + for making it easier to control opacity with an expression rather than having to do this separately for every property.

**Comment by nickidlugash** [comment]
I'm not sure that adding a linear `hillshade-opacity` property would be that helpful, because in many cases the reduced range would blow out important details in the hillshading.

`hillshade-exaggeration` is essentially a non-linear opacity property that takes into account the variables that @mollymerp mentioned. The combination of the property name and value range (0–1, with 0.5 being the default) though is perhaps confusing...`hillshade-exaggeration: 0.5` is actually _no_ exaggeration. >0.5, the proportion of opaque pixels increases, while <0.5, all pixels become increasingly less opaque. In the shader itself, we call this property `intensity` – maybe this would be a better style property name? We could also just rename it `hillshade-opacity` (although I don't particularly like that option since this property doesn't behave the same way as other opacity properties). 

@stevage would `hillshade-exaggeration` work for your use cases? Your feedback would be appreciated! FWIW, I think adjusting alpha values for the three color properties separately is also valuable in some cases, since you might want to adjust the balance between shadow, highlight, and slope.

**Comment by stevage** [comment]
Hmmm, interesting comments about `hillshade-exaggeration`. I did find, that playing with it, it didn't work as I imagined - thought it would be like the exaggeration parameter that you can pass to gdal's hillshade tool (whose name I've forgotten). So I was expecting to be able to increase the parameter without limit to get even more extreme shading.

So, no, I wouldn't rename that one `-opacity`, and I think `-intensity` makes the most sense.

> so if we did allow a hillshade-opacity property, it would behave differently than the -opacity properties on other layers.

I didn't quite understand how this follows, @mollymerp. Are you saying it's not possible to apply a transform to the alpha channel after all other calculations (including lighting etc)? IMHO it's not inherently confusing that opacity is determined by several sources (eg, a color property that has an alpha channel, plus an overall opacity). That's already the case with textures that contain alpha channels too, right?

If an `-opacity` property is not going to be supported, consider how these two use cases would be supported:

1. No hillshading at all below zoom 10, regular hillshading above.
2. No hillshading below zoom 10, increasingly visible hilshading to zoom  12, then fading out completely to 14.

(This latter use case is pretty similar to what I had on cycletour.org, using tilemill. I found it was distracting at low zooms, and too pixellated and not useful at high zooms.)

**Comment by nickidlugash** [comment]
>  thought it would be like the exaggeration parameter that you can pass to gdal's hillshade tool (whose name I've forgotten).

@stevage Yeah, this name _does_ make it sound like the property is an arbitrary z-factor that scales the calculated slope values. It is not, but values 0.5–1.0 are similar to a (limited) exaggeration factor that is capped to ensure that the hillshading displays reasonably well in all regions. We also do already include scaling in the shaders to adjust for both zoom level and projection distortion. Do you have use cases for an arbitrary z-factor that can't be addressed with the current `-exaggeration` (aka intensity) property?

> If an -opacity property is not going to be supported, consider how these two use cases would be supported:
> 1. No hillshading at all below zoom 10, regular hillshading above.
> 2. No hillshading below zoom 10, increasingly visible hilshading to zoom 12, then fading out completely to 14.

1. A `minzoom: 10` assignment in the style layer.
2. `minzoom: 10` and `maxzoom: 14`, plus an interpolate expression value for exaggeration (intensity), e.g.: 
`"hillshade-exaggeration": ["interpolate", ["linear"], ["zoom"], 10, 0, 12, 0.5, 14, 0 ]`

For values 0–0.5, `hillshade-exaggeration` is pretty similar to a uniform alpha transform with range 0–1  (it's not uniform, but the results aren't that different).

So this property is pretty similar to a combination of "opacity" and "exaggeration". To me it made sense at the time to combine them because the original, non-alpha-adjusted, non-exaggerated opacity level is not inherently more accurate or meaningful than any adjusted level, so I thought that all levels of opacity should just be on a single scale of 0–1, where 0=no opacity, and 1=full opacity (where "full" is the capped upper limit described above). I recognize that this deviates from how some other tools define their parameters. If this property doesn't work well for your use case, can you provide more details about the visual results you're looking to achieve? Or if it does work but the property name/value/default value seem confusing, can you describe how you would expect this property to be specified? Thanks so much for taking the time to provide feedback!

**Comment by stevage** [comment]
>Do you have use cases for an arbitrary z-factor that can't be addressed with the current -exaggeration (aka intensity) property?

I'm not sure. Most likely it would be if I really really want the hillshading to be visible despite a lot of other stuff being visible - some kind of temporary sacrificing of aesthetics for functionality. Or maybe for some kind of unusual visual effect, like when I was playing with having the direction of the sun rotating, I would have preferred to have that effect even stronger maybe.

>A minzoom: 10 assignment in the style layer.

I'm a bit wary of using `minzoom` in this way after recent experiences (see #5918) but perhaps that's unjustified.

>If this property doesn't work well for your use case, can you provide more details about the visual results you're looking to achieve? Or if it does work but the property name/value/default value seem confusing, can you describe how you would expect this property to be specified?

You ask very good questions :) I think I'll really have to play more, and try to create the style I want, and give more feedback then. I do feel that "intensity" is a more comfortable adjective than "exaggeration" for the current behaviour though (or at least, an "exaggeration" with an upper bound of 1 being counter-intuitive). Or even `hillshade-strength`?
